app/utils/array_utils.rb

```diff
@@ -46,4 +46,11 @@ def each_slice_columns(xs, columns, &block)
   def trim(xs)
     xs.drop_while { |x| x.blank? }.reverse.drop_while { |x| x.blank? }.reverse
   end
+
+  def zip_by(a, b, &block)
+    a.map { |a_elem|
+      b_found = b.find { |b_elem| block.call(a_elem, b_elem) }
```

Comment:
When the zipping condition is arbitrary there's no guarantee that only one element in array b matches with an array a element. I think I would expect this to handle that case too. You could do this by using reduce instead of map and find_all  + concatenation instead of find.

Another case is what if a b element matches with multiple a elements. Should it be duplicated or zipped into the first place where it matches? I'd say put it into first place and have no dups.